File: Cuarto/SO/Prácticas/P3/practica3/04-io-bound/ult.c
#include <pth.h>
#include <stdio.h>
#include <unistd.h>
#include <time.h>
#include "common.h" // clockwatch_*

#define SLEEP_TIME 10
#define NUM_THREADS 5

void* do_work(void* arg) {
    unsigned long thread_id = (unsigned long) pth_self();
    clockwatch_t clock;
    double elapsed;

    printf("[Thread %lu] Doing some work...\n", thread_id);
    clockwatch_start(&clock);

    pth_sleep(SLEEP_TIME); // Simulate some io-bound work

    elapsed = clockwatch_stop(&clock);
    printf("[Thread %lu] Done with work in %f seconds.\n", thread_id, elapsed);
    return NULL;
}

int main() {
    pth_t threads[NUM_THREADS];
    clockwatch_t clock;
    double elapsed;

    pth_init();

    printf("Starting the program.\n");
    clockwatch_start(&clock);

    for (int i = 0; i < NUM_THREADS; i++) {
        threads[i] = pth_spawn(PTH_ATTR_DEFAULT, do_work, NULL);
    }

    for (int i = 0; i < NUM_THREADS; i++) {
        pth_join(threads[i], NULL);
    }

    elapsed = clockwatch_stop(&clock);
    printf("All threads are done in %f seconds\n", elapsed);

    pth_kill();

    return 0;
}

File: Cuarto/SO/Prácticas/P3/practica3/04-io-bound/ult.py
import gevent
import time

# Delay in seconds to simulate IO-bound work
# adjust for faster results or to have extra time to monitor resources
SLEEP_TIME = 10

def do_work():
    greenlet_id = id(gevent.getcurrent())
    print(f"[{greenlet_id=}] Doing some work...")
    gevent.sleep(SLEEP_TIME)
    print(f"[{greenlet_id=}] Done with work.")


if __name__ == "__main__":
    start = time.time()
    print("Starting the program.")
    all_greenlets = []
    for i in range(5):
        # Start 5 greenlets
        greenlet = gevent.spawn(do_work)
        all_greenlets.append(greenlet)

    gevent.joinall(all_greenlets)
    elapsed = time.time() - start
    print(f"All greenlets are done in {elapsed} seconds")